src/query/api/v1/handler/graphite/find.go

```diff
@@ -57,52 +59,94 @@ func NewFindHandler(
 	}
 }
 
+func mergeTags(
+	terminatedResult *storage.CompleteTagsResult,
+	childResult *storage.CompleteTagsResult,
+) (map[string]bool, error) {
+	// sanity check the case.
+	if terminatedResult.CompleteNameOnly {
+		return nil, errors.New("terminated result is completing name only")
+	}
+
+	if childResult.CompleteNameOnly {
+		return nil, errors.New("child result is completing name only")
+	}
+
+	mapLength := len(terminatedResult.CompletedTags) + len(childResult.CompletedTags)
+	tagMap := make(map[string]bool, mapLength)
+
+	for _, tag := range terminatedResult.CompletedTags {
+		for _, value := range tag.Values {
+			tagMap[string(value)] = false
+		}
+	}
+
+	// NB: fine to overwrite any tags which were present in the `terminatedResult` map
+	// since if they appear in `childResult`, then they exist AND have children.
+	for _, tag := range childResult.CompletedTags {
+		for _, value := range tag.Values {
+			tagMap[string(value)] = true
+		}
+	}
+
+	return tagMap, nil
+}
+
 func (h *grahiteFindHandler) ServeHTTP(
 	w http.ResponseWriter,
 	r *http.Request,
 ) {
 	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
 	logger := logging.WithContext(ctx)
 	w.Header().Set("Content-Type", "application/json")
-	query, rErr := parseFindParamsToQuery(r)
+
+	// NB: need to run two separate queries, one of which will match only the
+	// provided matchers, and one which will match the provided matchers with at
+	// least one more child node. For further information, refer to the comment
+	// for parseFindParamsToQueries
+	terminatedQuery, childQuery, raw, rErr := parseFindParamsToQueries(r)
 	if rErr != nil {
 		xhttp.Error(w, rErr.Inner(), rErr.Code())
 		return
 	}
 
-	opts := storage.NewFetchOptions()
-	result, err := h.storage.FetchTags(ctx, query, opts)
-	if err != nil {
+	var (
+		terminatedResult *storage.CompleteTagsResult
+		tErr             error
+		childResult      *storage.CompleteTagsResult
+		cErr             error
```

Comment:
nit: You could make this (to slightly compact):
```go
var (
  terminateResult, childResult *storage.CompleteTagsResult
  termErr, childErr error
  // ...
)
```

Reply:
Actually not a fan of doing it like that, looks messier to me :(

If it's a standard we're using will update though :+1: